### DISCUSSION TITLE
How to use hyperloglog column outside presto?

### DISCUSSION THREAD

**Discussion by johnchenghk01** [prestodb/presto#14630] [General]
Is it able to read hyperloglog(stored in varbinary) with python?

**Comment by qqwhy666** [comment]
This question centers on using Presto's HyperLogLog data outside of the Presto environment (specifically in Python), and we can break it down by first clarifying key concepts:
Key Concept Background
HyperLogLog: An efficient algorithm for approximating the number of unique elements (cardinality) in a dataset (e.g., counting website UVs). It uses minimal memory, making it ideal for big data scenarios.
Presto (now often called Trino): A distributed SQL query engine. It includes a built-in HyperLogLog column type—you can use the HyperLogLog algorithm to compute cardinality directly in Presto, then store the resulting calculation state as varbinary (a binary data format) in backend storage (e.g., Hive, S3).
varbinary: A binary data type; here, it’s the format Presto uses to encode HyperLogLog’s computed state into a binary string for storage.
Core Goal of the QuestionThe asker’s scenario is this: They’ve already created HyperLogLog columns in Presto, and this data is stored as varbinary in their storage system. Now, they want to work outside of Presto—specifically, use Python to read this varbinary-encoded HyperLogLog data (i.e., parse the binary string in Python to retrieve the cardinality (unique count) or other stats captured by the HyperLogLog structure).
Key Challenge HereDifferent systems (including Presto) use custom, system-specific encoding rules for HyperLogLog’s binary format (there’s no universal standard for this encoding). So, to read Presto’s HyperLogLog varbinary data in Python, you’d need to reference or reverse-engineer Presto’s internal logic for encoding HyperLogLog state into varbinary—generic Python HyperLogLog libraries can’t parse it directly, as they don’t align with Presto’s custom binary structure.
